internal/uvm/virtual_device.go

```diff
@@ -0,0 +1,39 @@
+package uvm
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	resourcePathVpciFmt = "VirtualMachine/Devices/VirtualPci/%s"
+)
+
+func (uvm *UtilityVM) GpuCounter() uint64 {
```

Comment:
It seems like this is just a convenience for the caller since we never use this function or `gpuCounter` internally. In that case, it shouldn't be part of `UtilityVM`.

Reply:
The goal of the gpuCounter is to keep track of the number of nvidia devices on a uvm so we can forward the device index to opengcs to pass through to the nvidia cli tool.

Reply:
Right but why is this part of the `UtilityVM` struct? Why can't the caller manage this state?

Also how do we know that the device index will be what we expect? If we remove a GPU and add it back, will it get a new index?

Reply:
I'm not sure what you're referring to specifically by caller here? The UtilityVM struct maintains other device information, I figured tracking the gpu devices on it would be appropriate. 

Right, good point, that needs to be looked into. Haven't finished testing out all scenarios yet :)

Reply:
This has been resolved with the comment [here](https://github.com/microsoft/hcsshim/pull/765#discussion_r376011544)